Reland basic dynamo support for traceable collectives

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100476

Relative to the original land, this also contains:
- Fix torchdeploy import of functional collectives
- Can't import torchdynamo utils due to torch._refs being missing

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

wconstab committed:
Fix torchdeploy import of functional collectives

Can't import torchdynamo utils due to torch._refs being missing

[ghstack-poisoned]

kumpera reviewed (approved):
OMG. Nice fix though.

wconstab committed:
Update on "Fix torchdeploy import of functional collectives"


Can't import torchdynamo utils due to torch._refs being missing

[ghstack-poisoned]

wconstab committed:
Update on "Fix torchdeploy import of functional collectives"


Can't import torchdynamo utils due to torch._refs being missing

[ghstack-poisoned]

wconstab committed:
Update on "Reland basic dynamo support for traceable collectives"


Relative to the original land, this also contains:
- Fix torchdeploy import of functional collectives
- Can't import torchdynamo utils due to torch._refs being missing

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge -ic

wconstab commented:
@pytorchbot merge -f"flaky infra"
